src/core/mount.c

```diff
@@ -1883,7 +1885,18 @@ static int mount_load_proc_self_mountinfo(Manager *m, bool set_flags) {
                 if (!device || !path)
                         continue;
 
-                device_found_node(m, device, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MOUNT);
+                d = strdup(device);
+                if (!d)
+                        continue;
+
+                /* Just to achieve device name uniqueness using a hashmap, without any practical
+                 * significance of the value. */
+                r = hashmap_ensure_put(&devices, &path_hash_ops_free, d, d);
```

Comment:
You can use `set_put_strdup_full()`

Reply:
This only uses hashmap to achieve duplicate removal of device names. Since we will not iterate over this table later, I think it's enough to use strdup(). set_put_strdup_full() seems redundant.

Reply:
I mean, `Hashmap` is redundant. Please use `Set`.

Reply:
Depending on @yuwata, you might need to implement code like the following instead of using `hashmap_ensure_put`:
```
_cleanup_set_free_ Set * devices；

if (!devices || !set_contains(devices, d)) {
    r = set_put_strdup_full(&devices, &path_hash_ops_free, d);
    if (r < 0)
        ....

    device_found_node(m, device, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MOUNT);
}
```

Reply:
@JetXujing what @yuwata is saying is that you should use the `Set` data structure not the `Hashmap` data structure. They are internally the same mostly, but if you have a hashmap where key and value are the same object then `Set` is the way to do. 

set_put_strdup_full() is really the way to go here, it simplifies this code greatly, just do:

```c
/* If the same device is listed twice, only process it once. Note that the suppresion of the duplicate processing is merely an optimization, hence in case of OOM (unlikely) we'll just process it twice. */
if (set_put_strdup(&full, &path_hash_ops_free, d) != -EEXIST)
        device_found_node(m, device, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MOUNT);
```

That reduces the whole patch to basically 3 lines of codes added. Isn't that nice?

Reply:
ah, right, so it becomes this:

```c
/* If the same device is listed twice, only process it once. Note that the suppresion of the duplicate processing is merely an optimization, hence in case of OOM (unlikely) we'll just process it twice. */
if (set_put_strdup(&full, &path_hash_ops_free, d) != 0)
        device_found_node(m, device, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MOUNT);
```

Reply:
Thanks, @poettering @yuwata @JetXujing , 3 lines of concise code are presented.